### ISSUE TITLE
Cannot resume training using model.save and load_model()

### ISSUE THREAD

**Issue by suraj-maniyar** [tensorflow/tensorflow#40342]
**System Information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
Yes

- OS Platform and Distribution:
CentOS Linux 7.6.1810

- TensorFlow installed from : 
binary (pip)

- TensorFlow version :
2.2.0-rc3

- Python version:
3.6.4

- CUDA/cuDNN version:
CUDA 10.1 / cuDNN 7.6.0

- GPU model and memory:
2 x TitanX - 12GB

**Describe the current behavior**
I am using Tensorflow 2.2.0 on multi-gpu system. Having the need to train large networks for several days, I save the model weights with optimizer state using ```model.save()```. When I reload the model using ```tf.keras.models.load_model()```, the loss spikes sharply on TensorBoard and the accuracy also shows a sudden drop. Though the loss recovers within the epoch, it does not comply with the intended behavior of saving training state using ```model.save()```.

**Describe the expected behavior**
The API should be able to save and resume training from the very same point after loading a model from '.h5' file.

**Standalone code to reproduce the issue**
This code is a minimal reproducible example. It was tested on multi-gpu systems with 8 gpus. The re-run of the script is achieved by deleting the current model and distribute strategy and re-initializing them to simulate stop and restart of training process.
```
import os
import glob
import numpy as np
import tensorflow as tf
tf.__version__

gpus = tf.config.experimental.list_logical_devices('GPU')
print(gpus)

RESULT_DIR = os.path.join(os.getcwd(), 'Test', 'Results')
CHECKPOINT_FREQUENCY = 16
LOG_EVERY = 1

BATCH_SIZE_PER_GPU = 16
NUM_GPUS = len(gpus)
GLOBAL_BATCH_SIZE = BATCH_SIZE_PER_GPU * NUM_GPUS

def get_model():
    
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(filters=32, strides=1, kernel_size=(4,4), input_shape=(28,28,1)),
        tf.keras.layers.Activation('relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(10)
    ])
    
    return model

class SparseCategoricalLoss(tf.keras.losses.Loss):
    
    def __init__(self, num_classes, name='SparseCategoricalLoss', from_logits=False, loss_weight=1.0, *args, **kwargs):
        
        super().__init__(*args, **kwargs)
        self.num_classes = num_classes
        self.name = name
        self.from_logits=from_logits
        self.loss_weight = loss_weight
        
    def loss_fn(self, y_true, y_pred):
        label = y_true[:,0:self.num_classes]
        logit = y_pred[:,0:self.num_classes]
        loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=self.from_logits,
                                                             name=self.name,
                                                             reduction=tf.keras.losses.Reduction.NONE)(label, logit)
        loss *= self.loss_weight
        return loss
    
    
    def call(self, y_true, y_pred):
        total_loss = self.loss_fn(y_true, y_pred)
        return total_loss

    def get_config(self):
         
        config = super().get_config().copy()
        config.update({
            'num_classes' : self.num_classes,
            'name' : self.name,
            'loss_weight' : self.loss_weight
        })
        return config

loss = SparseCategoricalLoss(num_classes=10,
                             from_logits=True,
                             name='categorical_loss')

strategy = tf.distribute.MirroredStrategy()

with strategy.scope():
    
    model = get_model()
    
    optimizer = tf.keras.optimizers.RMSprop(
                                            learning_rate=0.001,
                                            epsilon=1.0,
                                            momentum=0.9,
                                            rho=0.9
                                           )
    
    model.compile(optimizer=optimizer, loss=loss, metrics=['acc'])

(X_train, Y_train), (X_test, Y_test) = tf.keras.datasets.mnist.load_data()
X_train = np.expand_dims(X_train, 3)
X_test = np.expand_dims(X_test, 3)

class LoggingCallback(tf.keras.callbacks.Callback):

    def __init__(self, result_dir, log_every, initial_step=0, checkpoint_frequency=None, **kwargs):
        
        super().__init__(**kwargs)
        
        # Create result directory
        self.result_dir = result_dir
        if not os.path.exists(result_dir):
            os.makedirs(result_dir)
        
        # create checkpoint directory
        checkpoint_dir = os.path.join(self.result_dir, 'checkpoint')
        if not os.path.exists(checkpoint_dir):
            os.makedirs(checkpoint_dir)
        
        # create tensorboard directory
        tensorboard_dir = os.path.join(self.result_dir, 'tensorboard')
        if not os.path.join(tensorboard_dir):
            os.makedirs(tensorboard_dir)
        
        self.log_every = log_every
        self.checkpoint_frequency = checkpoint_frequency
        self.train_writer = tf.summary.create_file_writer( os.path.join(tensorboard_dir, 'train') )
        self.step = initial_step
        
        
    # Write metrics to TensorBoard    
    def write_metrics_tensorboard(self, logs):
        with self.train_writer.as_default():
            for name, value in logs.items():
                if name in ['batch', 'size']:
                    continue
                tf.summary.scalar(name, value, step=self.step)
                
                
    def on_batch_end(self, batch, logs=None):
        
        self.step += 1
        
        # Write metrics to tensorboard
        if self.step % self.log_every == 0:
            self.write_metrics_tensorboard(logs)
            
        # Save model checkpoint (weights + optimizer state)
        if self.checkpoint_frequency and self.step % self.checkpoint_frequency == 0:
            name = 'model_step_%d.h5' % self.step
            path = os.path.join(self.result_dir, 'checkpoint', name)
            self.model.save( path )

callbacks = LoggingCallback(result_dir=RESULT_DIR, log_every=LOG_EVERY, checkpoint_frequency=CHECKPOINT_FREQUENCY)

model.fit(
          x = X_train, 
          y = Y_train, 
          batch_size=GLOBAL_BATCH_SIZE,
          epochs=7,
          validation_data = (X_test, Y_test),
          callbacks=callbacks,
          verbose=1 
         )

del model
del strategy

previous_checkpoints = glob.glob(os.path.join(RESULT_DIR, 'checkpoint', '*'))
previous_checkpoints.sort(key=lambda x : int(os.path.basename(x).split('_')[2].replace('.h5', '')) )
latest_checkpoint = previous_checkpoints[-1]
print('Found Latest Checkpoint : %s' % latest_checkpoint)
    
initial_step = int(os.path.basename(latest_checkpoint).split('_')[2].replace('.h5', ''))
print('Resuming training from step %d' % initial_step)
    
new_callback = LoggingCallback(result_dir=RESULT_DIR, log_every=LOG_EVERY, initial_step=initial_step, checkpoint_frequency=CHECKPOINT_FREQUENCY)

strategy = tf.distribute.MirroredStrategy()
with strategy.scope():
    model = tf.keras.models.load_model( latest_checkpoint, custom_objects={'SparseCategoricalLoss':SparseCategoricalLoss} )

model.fit(
          x = X_train, 
          y = Y_train, 
          batch_size=GLOBAL_BATCH_SIZE,
          epochs=10,
          validation_data = (X_test, Y_test),
          callbacks=new_callback,
          verbose=1 
         )

```

Here is a link to colab showing the output : https://colab.research.google.com/gist/suraj-maniyar/1a305d7249baee4393147cb479ea2933/restart_training.ipynb

**Other info / logs** 

The TensorBoard entry looks like this : 
![tensorboard](https://user-images.githubusercontent.com/16310456/84227046-66de4f00-aab1-11ea-86eb-9bcb6de2aa8d.PNG)

This was a toy example using mnist. After about 26k steps, when the training was restarted, the loss spiked up indicating that the last saved checkpoint did not save the training configuration correctly.
I am training an InceptionResNet network for several days and the spike in the loss is very concerning when I restart the training (shown below).
![tensorboard_inception](https://user-images.githubusercontent.com/16310456/84227678-04864e00-aab3-11ea-940d-85748d4823ec.PNG)

**Comment by suraj-maniyar** [comment]
@anj-s 
I tried saving the model using "tf" format and  the problem still persists.
See gist [here](https://colab.research.google.com/gist/suraj-maniyar/048b5697090e7ffc1bf09ca14fb8da01/copy-of-restart_training_tf_save.ipynb). 
Any work-around for this issue?

**Comment by rchao** [comment]
@suraj-maniyar question for you - if you do `model.fit()`, then save, and load immediately, does the loaded model retain the weight that was just saved?

**Comment by suraj-maniyar** [comment]
@rchao Yes the weights are same when I load the model right after ```model.fit()``` and ```model.save()```.
But the training after loading the weights still shows the spike in loss and gets worse from there on. ([gist](https://colab.research.google.com/gist/suraj-maniyar/a7c4b8009bc71f951204eb40018706ab/copy-of-restart_training_tf_save_v2.ipynb))

**Comment by rchao** [comment]
So to be clear, if you save and load, the weights are the same, but if you continue training, the loaded model would have incorrect loss and accuracy from the original model you saved from?

**Comment by suraj-maniyar** [comment]
Yes. Precisely.

**Comment by suraj-maniyar** [comment]
@rchao Any updates?

**Comment by suraj-maniyar** [comment]
@nikitamaia @anj-s @rchao 
Are there any updates on this issue?

**Comment by rchao** [comment]
Sorry for the late response - but based on the information here I wasn't able to tell where went wrong. Will reply here if we have any updates.

**Comment by w4nderlust** [comment]
Could it be the batch normalization loosing history of the momentum?
Try without the batchnorm layer and see if you still see the spikes.

**Comment by suraj-maniyar** [comment]
@w4nderlust 
I still see the problem even after removing batch normalization ([gist](https://colab.research.google.com/gist/suraj-maniyar/e3bd058be7efafc1c3a34456d3b85ebd/copy-of-restart_training_tf_save_v2.ipynb)).

**Comment by w4nderlust** [comment]
> @w4nderlust
> I still see the problem even after removing batch normalization ([gist](https://colab.research.google.com/gist/suraj-maniyar/e3bd058be7efafc1c3a34456d3b85ebd/copy-of-restart_training_tf_save_v2.ipynb)).

If batch norm is not the problem, it could be that this problem is related to this issue I opened yesterday: https://github.com/tensorflow/tensorflow/issues/41053
In that case, my problem is that after resuming a stateful optimizer in a custom training loop (by means of pickling and unpickling) the learning curves differ.
In my case, if I use a stateless optimizer, like SGD, I see no problem, the problems only appear when using a stateful optimizer like Adam.
Maybe, just to confirm if that is the problem, you could test your code using SGD instead of RMSProp. If you don't see the problem, then it's likely that the issue is with the resume logic of the optimizer.

**Comment by suraj-maniyar** [comment]
@w4nderlust 
I still see the issue with SGD optimizer ([gist](https://colab.research.google.com/gist/suraj-maniyar/71092e331e37d66fd92ecfcf027bc77c/copy-of-test1.ipynb)). 
I had tried something similar to what you mentioned in your issue i.e. save and load optimizer state separately as pickle object, but I was still seeing the spike in loss.
I wonder if ```tf.distribute.Strategy``` has a different logic to save and reload model and causing the optimizer to reset.

**Comment by guptapriya** [comment]
@suraj-maniyar apologies for the delay here. In your experience, does this issue show up even when not using any distribution strategy? I tried one of your [gists](https://colab.research.google.com/gist/suraj-maniyar/048b5697090e7ffc1bf09ca14fb8da01/copy-of-restart_training_tf_save.ipynb) and changed it from using MirroredStrategy to the [default strategy](https://www.tensorflow.org/guide/distributed_training#default_strategy) - using default strategy is equivalent to not using any distribution strategy. And I see the same effect on loss. 

If this is the case, then perhaps this is a more general issue and not specific to dist strat, and Keras team maybe in best position to help debug. My best guess would be something weird is going on with saving/loading the custom loss object. @pavithrasv maybe able to help look, as I don't have much experience with saving/loading these custom objects in keras.

**Comment by suraj-maniyar** [comment]
Yes this issue is persistent even without any distribution strategy. Here's my [gist](https://colab.research.google.com/gist/suraj-maniyar/4539773fdaa9a3a8ab6be0de5a745e1b/resume_train_no_dist_strat.ipynb) if you need to look @guptapriya @pavithrasv

**Comment by guptapriya** [comment]
Thanks for confirming @suraj-maniyar. We will ask the folks who are expert in Keras to take a look.

**Comment by vladbph** [comment]
Same issue here as well. Tried TF2.0 and 2.2 - same thing. Quite basic scenario:
1. Train model for x epoch
2. Save model as json and weights separately
3. Restart process to resume training:
    - load model from json
    - load weights
    - model.fit()
Expected the 'loss' to be close the last best saved value (monitor = 'loss'). Instead, model starts training like weights were not loaded at all. Strategy does not affect the result.

**Comment by vladbph** [comment]
> Same issue here as well. Tried TF2.0 and 2.2 - same thing. Quite basic scenario:
> 
> 1. Train model for x epoch
> 2. Save model as json and weights separately
> 3. Restart process to resume training:
>    
>    * load model from json
>    * load weights
>    * model.fit()
>      Expected the 'loss' to be close the last best saved value (monitor = 'loss'). Instead, model starts training like weights were not loaded at all. Strategy does not affect the result.
> 
> Note. When all of the above steps are run sequentially in a process(without restart - step #3) model will continue training as expected.

Digging further: 
- Model that uses **_binary_crossentropy_** works **OK** 
- Model that uses **_categorical_crossentropy_** **FAILs** to continue training from last checkpoint

**Comment by vladbph** [comment]
This may be sounds too far fetched, but it seems that ..... python version is the reason. When model is compiled with **_categorical_crossentropy_**:

python2.7:
	*** categorical_crossentropy (tf1.4.0 + keras2.1.5) // OK
	*** categorical_crossentropy (tf2.0.0) // OK
	*** categorical_crossentropy (tf2.1.0) // OK

python 3.6:
	*** categorical_crossentropy (tf1.4.0 + keras2.1.5) // FAILS
	*** categorical_crossentropy (tf2.1.0) // FAILS
	*** categorical_crossentropy (tf2.2.0) // FAILS

Loading weights only or whole model  - there is no effect on the results. However **_binary_crossentropy_** works for all cases above. I did not tested other loss functions though...

**Comment by Lapiquettejack** [comment]
I had the same problem than suraj-maniyar on a custom training with custom loss. I'm using a stateful optimizer as well but without any distribution strategy. I found a satisfying solution here : https://stackoverflow.com/questions/49503748/save-and-load-model-optimizer-state. 

The problem was to resume the optimizer state, which was possible without compiling the model, this solution works with Adam could be adapted to other optimizer I guess.

Hope this helps.

**Comment by suraj-maniyar** [comment]
@Lapiquettejack 
I have tried that and it didn't work for me.

**Comment by yhliang2018** [comment]
Hi @suraj-maniyar, based on your last [gist](https://colab.sandbox.google.com/gist/suraj-maniyar/4539773fdaa9a3a8ab6be0de5a745e1b/resume_train_no_dist_strat.ipynb) link, I've tried to add `model.compile(optimizer=optimizer, loss=loss, metrics=['acc'])` right after the model is loaded via `tf.keras.models.load_model`. The training accuracy of the loaded model is improved a lot compared to that in your gist. 

Unfortunately, the Tensorboard was not launched successfully on my end. Could you try the same to see how the Tensorboard metrics looks like?

**Comment by byronyi** [comment]
Echo with a similar issue https://github.com/keras-team/keras/issues/14231

**Comment by thecosta** [comment]
Same issue here:
1. trained model in TF==2.2.0. Saved in SavedModel format, and JSON and .h5. 
2. from new session, load model and continue/resume training. Out of the box, the accuracy starts from 0 and loss is high.

**Comment by PeteAndrews** [comment]
> Could it be the batch normalization losing history of the momentum?
> Try without the batchnorm layer and see if you still see the spikes.

Same issue! I found it is batchnorm momentum causing the problem, as you suggested. Checkpoints were not saving my optimiser (Adam) attributes, so I exported these separately. However, I am not sure how to access the batchnorm layer attributes. Any suggestions would be extremely helpful ;)

**Comment by evanatyourservice** [comment]
Still happens in 2.6.0. I read elsewhere this happens with distributed strategies. Optimizer weights are not saved in the savedmodel, and so are not loaded when restarting from the savedmodel. Nowhere is the tf docs does it warn about this. I don't mind if it's the case, but at least update the docs and give a workaround, this has been an issue for 2+ years.

My workaround steps are:

1. save two checkpoints during training, one h5 model checkpoint, one weights only checkpoint
2. when resuming, load the model from the h5, then model.load_weights() from the weights only checkpoint
3. model.save("path", include_optimizer=False) after training to create a savedmodel for inference

You can print the model.optimizer.weights after this. If you try to load a savedmodel saved during distributed training you cannot print model.optimizer.weights because they don't exist. And if you try to load a weights checkpoint into a loaded savedmodel, it will say the optimizer weights are unused.

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
